Add the Ability to filter file uploads based on file extensions

Sigmonia opened the pull request:
#### Rationale
[Secure Issue 51524: Allowlist for acceptable types of file upload](https://www.labkey.org/home/Developer/issues/Secure/issues-details.view?issueId=51524)

#### Related PRs
* https://github.com/LabKey/premiumModules/pull/100

#### Changes
- updated shared jsps
- Add allow list
- and extension checker to file uploads

Sigmonia committed:
Add the ability to limit file uploads based on the file extension

labkey-jeckels commented on api/src/org/labkey/api/util/FileUtil.java:
```diff
@@ -316,9 +318,33 @@ public static String isAllowedFileName(String s)
+    public static void setExtensionChecker()
+    {
+        // Regex encode the allowed extensions (escape periods and add '|' optional matcher)
+        String allowedExtensions = String.join("|", AppProps.getInstance().getAllowedExtensions()).replace(".", "\\.");
```
What about other special regex characters, like `-` or `(`?

This seems like a good candidate for a small unit test.

Sigmonia replied:
test added, using Pattern.quote()

labkey-jeckels commented on core/src/org/labkey/core/admin/ExternalServerType.java:
```diff
@@ -83,8 +94,60 @@ public void setHosts(Collection<String> hosts)
+            return HtmlString.unsafe("""
+                <div style="width: 700px">
+                    <div>
+                        This list is the set of file extensions that LabKey will allow through. Any extension that is not in this list will be rejected, this includes mulitple extensions. If the list is blank, then this check is ignored.
```
```suggestion
                        This list is the set of file extensions that LabKey will accept for uploads. Any extension that is not in this list will be rejected, this includes multiple extensions. For example, .gz is not sufficient to allow .tar.gz; you must specify .tar.gz. If the list is blank, then this check is ignored.
```

Sigmonia replied:
updated

labkey-jeckels commented on api/src/org/labkey/api/util/FileUtil.java:
```diff
@@ -316,9 +318,33 @@ public static String isAllowedFileName(String s)
         if (Pattern.matches("(.*\\s--[^ ].*)|(.*\\s-[^- ].*)",s))
             return "Filename may not contain space followed by dash.";
+
+        String badExtension = checkExtension(s);
```
This method is only invoked if the server has been configured to block these patterns, via a separate property. It's also only consulted for WebDav uploads.

The customer who raised this concern would want it to apply to all uploads, including attachments to issues, wikis, etc.

labkey-jeckels replied:
I think that means also checking in AntiVirusMultipartResolver.

Sigmonia replied:
Added it to the AntiVirusMultipartResolver. Its unclear to me how to best add this call to the Spring `StandardServletMultipartResolver`

labkey-jeckels replied:
Looks like we'll need another subclass of `StandardServletMultipartResolver` that lives in the API module. The default implementation in AntiVirusProviderRegistry should return that new subclass, which looks at the file name. We can refactor `AntiVirusMultipartResolver` to inherit from the new subclass and reduce duplicated code.

labkey-jeckels commented on core/src/org/labkey/core/admin/AdminController.java:
```diff
@@ -449,6 +449,8 @@ public static void registerAdminConsoleLinks()
         AdminConsole.addLink(Configuration, "system maintenance", new ActionURL(ConfigureSystemMaintenanceAction.class, root));
         AdminConsole.addLink(Configuration, "external redirect hosts", new ActionURL(ExternalHostsAdminAction.class, root).addParameter("type", ExternalServerType.Redirect.name()), TroubleshooterPermission.class);
         AdminConsole.addLink(Configuration, "external allowed sources", new ActionURL(ExternalHostsAdminAction.class, root).addParameter("type", ExternalServerType.Source.name()), TroubleshooterPermission.class);
+        if (ModuleLoader.getInstance().hasModule("Premium"))
```
The combo of checking for the premium module and the class/enum names on the next line is too much for me.

We could clean it up by making this a non-premium feature. I'm fine with that given the implementation approach.

Sigmonia committed:
Add unit test
- fix regex quoting
- modified checker methods to be testable

adjusted description text

Sigmonia committed:
Make call from

labkey-jeckels commented on api/src/org/labkey/api/util/FileUtil.java:
```diff
@@ -2221,5 +2253,62 @@ public void testAllowedFileName()
+            assertNotNull("Matched unlist extension", checkExtension("my test.notListed", mockProps));
+            assertNotNull("Combined multiple extension matched incorrectly", checkExtension("multi.a_v.tar", mockProps));
+            assertNotNull("Multi-multi extension matched unexpectedly", checkExtension("multi.not.tar.gz", mockProps));
+            assertNotNull("No extension matched unexpectedly", checkExtension("No extension", mockProps));
```
I hadn't considered files with no extension. I guess it's OK to reject them all at this point.

labkey-jeckels commented on core/src/org/labkey/core/admin/ExternalServerType.java:
```diff
@@ -83,8 +100,72 @@ public void setHosts(Collection<String> hosts)
+            return HtmlString.unsafe("""
+                <div style="width: 700px">
+                    <div>
+                        This list is the set of file extensions that LabKey will accept for uploads. Any extension that is not in this list will be rejected, this includes multiple extensions. For example, .gz is not sufficient to all .tar.gz; you must specify .tar.gz. If the list is empty, then this check is ignored.
```
Typo in my suggestion: "is not sufficient to all" should be "is not sufficient to allow"

labkey-jeckels commented on api/src/org/labkey/api/util/FileUtil.java:
```diff
@@ -316,9 +324,33 @@ public static String isAllowedFileName(String s)
             return "Filename may not begin with any of these characters: -$";
```
We should pull the check of `AppProps.getInstance().isInvalidFilenameUploadBlocked()` out of `DavController.checkAllowedFileName()` and put it here around the prexisting set of checks. Otherwise, we're ignoring that setting.

labkey-jeckels commented on core/src/org/labkey/core/webdav/DavController.java:
```diff
@@ -3074,7 +3074,10 @@ public void closeInputStream() throws IOException
-
+            if (FileUtil.isAllowedFileName(name) != null)
+            {
+                throw new IOException("The file extension is not allowed.");
```
This may reject the name unrelated to the exception. Let's use the message this call returns instead of hard-coding the error message

Sigmonia committed:
CR feedback 2
- Move isInvalidFilenameBlocked check
- Subclass Multipart Resolver so we can more easily check the file extension
- fix typos and messaging

labkey-matthewb commented on api/src/org/labkey/api/util/FileUtil.java:
```diff
@@ -304,21 +312,48 @@ public static void copyDirectory(Path srcPath, Path destPath) throws IOException
+        if (appProps.getAllowedExtensions().isEmpty())
+            return null;
+
+        if (extensionChecker == null)
```
When does this get cleared?  It seems very sticky.

Sigmonia replied:
It only gets changed in setExtensionChecker which is called when the WritableAppProps.setAllowedFileExtensions is called when the new Allow List prop is saved or when the server is starting up

labkey-matthewb replied:
I see.  The pattern confused me.  I would have expected a getExtensionCheck() that wrapped the null check.

labkey-matthewb commented on core/src/org/labkey/core/webdav/DavController.java:
```diff
@@ -3075,6 +3075,11 @@ public void closeInputStream() throws IOException
+            String notAllowedMsg = FileUtil.isAllowedFileName(name);
+            if (StringUtils.isNotBlank(notAllowedMsg))
+            {
+                throw new IOException(notAllowedMsg);
```
throw DavException in  the dav controller to get the file browser to display an expected UI.  Maybe use 409 SC_CONFLICT?  or 406 (SC_NOT_ACCEPTABLE)

labkey-matthewb commented:
Do we want to allow folder names to have "." in them?  Disallowing doesn't seem to be related to the original goal of the feature.

Sigmonia committed:
rolling this back

Sigmonia committed:
Throwing a DevException to improve the error shown to the User

Sigmonia committed:
Don't check extensions for directories
Updated error message read the text passed

Sigmonia committed:
fix the unit test
